Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs

```diff
@@ -0,0 +1,207 @@
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            _appearanceComponent = Owner.GetComponent<AppearanceComponent>();
+        }
+
+        public override void OnAdd()
+        {
+            // This will blow up an entity it's attached to
+            base.OnAdd();
+            if (Owner.TryGetComponent<CollidableComponent>(out var collide))
+            {
+                collide.IsHardCollidable = false;
+            }
+
+            _state = PortalState.UnableToTeleport;
+            if (_aliveTime > 0)
+            {
+                Timer.Spawn(TimeSpan.FromSeconds(_aliveTime), () => Owner.Delete());
+            }
+        }
+
+        public override void OnRemove()
+        {
+            _appearanceComponent = null;
+
+            base.OnRemove();
+        }
+
+        public bool CanBeConnected()
+        {
+            if (_connectingTeleporter == null)
+            {
+                return true;
+            }
+            return false;
+        }
+
+        public void TryConnectPortal(IEntity otherPortal)
+        {
+            // TODO: Refactor to use out or w/e
+            if (otherPortal.TryGetComponent<ServerPortalComponent>(out var connectedPortal) && connectedPortal.CanBeConnected())
+            {
+                _connectingTeleporter = otherPortal;
+                connectedPortal._connectingTeleporter = Owner;
+                TryChangeState(PortalState.Pending);
+            }
+        }
+
+        public void TryChangeState(PortalState targetState)
+        {
+            if (Owner == null)
+            {
+                return;
+            }
+
+            _state = targetState;
+            if (_appearanceComponent != null)
+            {
+                _appearanceComponent.SetData(PortalVisuals.State, _state);
+            }
+        }
+
+        private void releaseCooldown(IEntity entity)
+        {
+            if (immuneEntities.Contains(entity))
+            {
+                immuneEntities.Remove(entity);
+            }
+
+            if (_connectingTeleporter != null &&
+                _connectingTeleporter.TryGetComponent<ServerPortalComponent>(out var otherPortal))
+            {
+                otherPortal.immuneEntities.Remove(entity);
+            }
+        }
+
+        private IEnumerable<IEntity> GetPortableEntities()
+        {
+            // TODO: Performance optimisation?
+            // TODO: If Teleportable put on a stored component (hands or storage) don't teleport it?
+            var portableEntities = new List<IEntity>();
+
+            foreach (var entity in _serverEntityManager.GetEntitiesIntersecting(Owner))
+            {
+                if (entity.HasComponent<TeleportableComponent>())
+                {
+                    portableEntities.Add(entity);
+                }
+            }
+
+            return portableEntities;
+        }
+
+        public void OnUpdate()
```

Comment:
I believe there should be events on `ICollidable` that get invoked when the entity starts overlapping. Use those and you won't have to expensively check every tick like this.

Reply:
I saw BumpedEntMsg under collidable so I could use that instead?

Reply:
Looking at it appears there actually isn't any such event. In that case just slap a TODO on it and keep it like this. Acruid was rewriting physics anyways soo..